packages/hir/src/lowering/types.candy

```diff
@@ -43,24 +43,39 @@ fun astTypeToHirType(
             .expand<HirDeclaration>({
               namePart.resolveIdentifier(context, it) as Iterable<HirDeclaration>
             })
-            .toSet() as Set<HirClass | HirTrait | HirProperty | HirFunction | HirModule>)
+            .toSet() as Set<HirType | HirTrait | HirFunction | HirModule>)
         }
 
         // There should only be one candidate left.
         let declaration = (candidates as Iterable<HirDeclaration>).single().unwrap()
-        assert((declaration is HirClass) || (declaration is HirTrait), "Declaration should be HirClass or HirTrait.")
-        return Tuple(HirNamedType(declaration as HirTrait | HirClass), List.empty<CompilerError>())
+        assert((declaration is HirType) || (declaration is HirTrait), "Declaration should be HirType or HirTrait.")
+
+        let types = ((astType as AstNamedType).typeArguments as Iterable<AstTypeArgument>)
+          .map<Maybe<HirInlineType>>({
+            it.type.map<HirInlineType>({ astInlineTypeToHirInlineType(context, it, scope) })
+          })
+        return Tuple(
+          if types.any({ it is None }) {
+            HirErrorType()
+          } else {
+            HirNamedType(
+              declaration as HirTrait | HirType,
+              types.map<HirInlineType>({ it.unwrap() }).toList(),
+            )
+          },
+          List.empty<CompilerError>(),
+        )
```

Comment:
We could return the actual named type and just mark specific type arguments as errors. (And with the next PR, the type arguments's `type` is no longer a `Maybe`.

```suggestion
        let types = ((astType as AstNamedType).typeArguments as Iterable<AstTypeArgument>)
          .map<HirInlineType>({
            it.type
              .map<HirInlineType>({ astInlineTypeToHirInlineType(context, it, scope) })
              .orElse({ HirErrorType() })
          })
          .toList()
        return Tuple(
          HirNamedType(declaration as HirTrait | HirType, types),
          List.empty<CompilerError>(),
        )
```